nomad/plan_apply.go

```diff
@@ -648,7 +648,7 @@ func evaluateNodePlan(snap *state.StateSnapshot, plan *structs.Plan, nodeID stri
 		return false, "node is not ready for placements", nil
 	} else if node.SchedulingEligibility == structs.NodeSchedulingIneligible {
 		return false, "node is not eligible for draining", nil
-	} else if node.Drain {
+	} else if node.DrainStrategy != nil {
```

Comment:
`node.Drain` and `node.DrainStrategy != nil` should be equivalent post 0.8.0, correct? so would that scenario only cause problem for upgrades directly from 0.8.0 to 1.1.0?

Reply:
> node.Drain and node.DrainStrategy != nil should be equivalent post 0.8.0, correct?

Correct! So your direct translation is *not wrong.* In >=0.8 it is merely redundant with the preceeding eligibility check.

> so would that scenario only cause problem for upgrades directly from 0.8.0 to 1.1.0?

Sadly no because even 1.0.x checks `node.Drain` here. So as soon as a server is upgraded to 1.1.0, the 1.0.x servers may improperly see `node.Drain = false` here due to the 1.1.0 server stripping the `true` value when roundtripping the Node struct.

But again! I _think_ we're ok in the `>=0.8 -> 1.1.0` upgrade case because of the SchedulingEligibility invariant and check above!

Upgrading from `<0.8` I'm not worried about. Our recommendation is not skipping versions, and skipping 1 or 2 almost always works fine. Skipping 6+ major releases is... not something I'm interested in supporting at this time. :sweat_smile:

Reply:
thanks, @schmichael , i was missing this nuance, worrying more about the newer versions and less about the older ones.

Reply:
we should consider these workflows when marking fields for deprecation in the future. adding an `IsDraining() bool` method at the time of deprecation might have eased this.